extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -59,34 +70,73 @@ private static QuarkusSecurityIdentity augmentIdentity(SecurityIdentity security
-            return authenticate(identityProviderManager, new IdTokenCredential(tokens[0], context))
-                    .thenCompose(new Function<SecurityIdentity, CompletionStage<SecurityIdentity>>() {
+            String idToken = tokens[0];
+            String accessToken = tokens[1];
+            String refreshToken = tokens[2];
+
+            return authenticate(identityProviderManager, new IdTokenCredential(idToken, context))
+                    .handle(new BiFunction<SecurityIdentity, Throwable, SecurityIdentity>() {
                         @Override
-                        public CompletionStage<SecurityIdentity> apply(SecurityIdentity securityIdentity) {
-                            return CompletableFuture
-                                    .completedFuture(augmentIdentity(securityIdentity, tokens[1], tokens[2], context));
+                        public SecurityIdentity apply(SecurityIdentity identity, Throwable throwable) {
+                            if (isLogout(context, configContext)) {
+                                // force logout, the challenge will be built accordingly and redirect to the logout endpoint at the OP
+                                throw new AuthenticationFailedException();
+                            }
+
+                            // if authentication failed, the reason could be the token no longer being valid, so we try a 
+                            // silent refresh if required
+                            if (throwable != null) {
+                                if ((identity = trySilentRefresh(configContext, idToken, refreshToken, context)) == null) {
+                                    // if the refresh fails, we just propagate the original exception
+                                    if (throwable instanceof AuthenticationFailedException) {
+                                        throw AuthenticationFailedException.class.cast(throwable);
+                                    }
+                                    throw new AuthenticationFailedException(throwable);
+                                }
+                            }
+
+                            return augmentIdentity(identity, accessToken, refreshToken, context);
                         }
                     });
         }
 
         // start a new session by starting the code flow dance
-        return performCodeFlow(identityProviderManager, context, resolver);
+        return performCodeFlow(identityProviderManager, context, configContext);
     }
 
     public CompletionStage<ChallengeData> getChallenge(RoutingContext context, DefaultTenantConfigResolver resolver) {
         TenantConfigContext configContext = resolver.resolve(context, false);
+        ServerCookie sessionCookie = (ServerCookie) context.cookieMap().get(SESSION_COOKIE_NAME);
+
+        if (isLogout(context, configContext)) {
+            String logoutPath = OAuth2AuthProviderImpl.class.cast(configContext.auth).getConfig().getLogoutPath();
```

Comment:
Might be worth adding a comment here it is an OIDC logout endpoint and check it is not null

Reply:
If a given OIDC does not support the logout, what should we return then ? The user is still logged in and does not need a challenge. I propose `LogoutException` (this actual exception will be in `quarkus-vertx-http`, as `AutenticationCompletionException` is) ? There we can do something about it, may be redirect to the logout.redirect_uri, in a timed way, return basic HTML page: "Logout is not supported, you will be redirect to {redirect.uri} in 3 seconds", something like that, as I don;t think 500 or 401 would be correct just because a user's attempt to logout is not supported :-)

Reply:
If the OP does not support logout we have some checks during the build so the application should fail to start.

Reply:
@pedroigor Ignore most of it please :-), I see a check further down the PR :-), Though what happens if the sessionCookie is null ?

Reply:
Yeah, I've added a check for this.

Reply:
@pedroigor so what happens if a logout has been requested and the session code is null ?

Reply:
Check added, as mentioned.

Reply:
@pedroigor It is just an NPE check :-) so what should happen to the logout request flow in this case ?

Reply:
You just go to the login page as per the normal flow.

Reply:
@pedroigor Got it :-)